src/ResourceManager/Common/Commands.ResourceManager.Common/Location/LocationCompleter.cs

```diff
@@ -0,0 +1,200 @@
+#endif
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            _resourceTypeLocationDictionary[contextHash] = new Dictionary<string, ICollection<string>>();
+#if DEBUG
+                            throw ex;
+#endif
+                        }
+                    }
+
+                    return _resourceTypeLocationDictionary[contextHash];
+                }
+            }
+        }
+
+        /// <summary>
+        /// Pass in a list of ResourceTypes and this class will provide a list of locations that are common to all ResourceTypes given. This will then be available to the user to tab through.
+        /// Example: [Parameter(ParameterSetName = ListByNameInTenantParameterSet, ValueFromPipelineByPropertyName = true, Mandatory = false), LocationCompleter(new string[] { "Microsoft.Batch/operationss" })]
+        /// </summary>
+        /// <param name="resourceTypes"></param>
+        public LocationCompleterAttribute(string[] resourceTypes) : base(CreateScriptBlock(resourceTypes))
+        {
+        }
+
+        public static string[] FindLocations(string[] resourceTypes)
+        {
+            return FindLocations(resourceTypes, ResourceTypeLocationDictionary);
+        }
+
+        /// <summary>
+        /// Finds all valid Locations for 
+        /// </summary>
+        /// <param name="resourceTypes"></param>
+        /// <param name="resourceTypeLocationDictionary"></param>
+        /// <returns></returns>
+        public static string[] FindLocations(string[] resourceTypes, IDictionary<string, ICollection<string>> resourceTypeLocationDictionary)
+        {
+            if (resourceTypeLocationDictionary == null)
+            {
+                resourceTypeLocationDictionary = new Dictionary<string, ICollection<string>>();
+            }
+
+            if (resourceTypes == null)
+            {
+                resourceTypes = new string[0];
+            }
+
+            List<string> validResourceTypes = new List<string>();
+            foreach (string resourceType in resourceTypes)
+            {
+                if (resourceType != null)
+                {
+                    if (resourceTypeLocationDictionary.ContainsKey(resourceType))
+                        validResourceTypes.Add(resourceType);
+#if DEBUG
+                    else throw new Exception(String.Format(Resources.InvalidResourceType, resourceType));
+#endif
+                }
+            }
+
+            string[] distinctLocations;
+            if (validResourceTypes.Count > 0)
+            {
+                distinctLocations = resourceTypeLocationDictionary[validResourceTypes[0]].ToArray();
+                foreach (string resourceType in validResourceTypes)
+                {
+                    distinctLocations = distinctLocations.Intersect(resourceTypeLocationDictionary[resourceType].ToArray(), new LocationEqualityComparer()).ToArray();
```

Comment:
what happens if distinctlocatiosn is null? (or is it always at least empty)?

Reply:
I added a check when creating validResourceTypes so that no resourceType with a null value in the dictionary will be added to validResourceTypes (I don't think this should ever be the case - right now it always creates an empty list but this is good in case the code changes in the future).